### ISSUE TITLE
binary classification inputs validation

### ISSUE THREAD

**Issue by stas00** [fastai/fastai#652]
Would it be sensible to add to `class MixedInputModel` the following validation?

```
if (is_reg==False):
    if is_multi==False: assert out_sz == 2, "for binary classification need out_sz=2"
    else:               assert out_sz > 2,  "for multiclass classification need out_sz>2"
```

it'd have saved me a lot of time trying to figure out why my code wasn't working as I didn't even think of checking `out_sz` which was set to 1, copied from the lesson's code (where it was regression). 

We don't really have a lesson notebook with binary classification for structured data, do we?

Also would it be easier for users if `out_sz` were called `n_classes`?

**Comment by sgugger** [comment]
I think you confuse is_multi with something else: it's like in lesson 2, the difference between the planet and the dogs/cats dataset. is_multi means we can have multiple labels for the same entry. Even with 2 outputs, it could be possible to have is_multi=True if the two outputs can be 1 together, and we can also have a model with n outputs where is_multi=False, it just means that you can only have one label that is true at a time.
It terms of internals, this just decides what is the proper loss last activation (a log_softmax when is_multi=False and a sigmoid when is_multi=True).

**Comment by sgugger** [comment]
Yup, though I'd prefer a warning because someone could design a classifier with just a 0/1 output (though they would have to be careful with the loss).

**Comment by stas00** [comment]
I'm not sure I follow you, this is exactly what I had the problem with - a binary classifier with 0/1 output. And after a lot of debug I finally figured it required out_sz=2, and then it all worked. Hence the suggestion. It doesn't work with out_sz=1 and bails with all kinds of mismatch issues on the CUDA side.

Are you talking about the same?

**Comment by sgugger** [comment]
Oh, disregard my comment. I had forgotten it was a model and not a learner wrapped around a model the user could define himself.

**Comment by stas00** [comment]
https://github.com/fastai/fastai/pull/654
